doc/design/context.md

```diff
@@ -0,0 +1,171 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run` to make computation. The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Operator::Run(OpContext* context);
+```
+
+And the `OpContext` is defined as follows:
+
+```c++
+Struct OpContext {
+  Scope* scope;
+  DeviceContext* device_context;
+};
+```
+
+The `Run` method will take `Variable` (containing `Tensor`) from `Scope` to make computation on certain `DeviceContext`. `DeviceContext` provides necessary runtime resources for computation, including CudaStream, cublasHandle and so on.
+
+Generally speaking, the Tensor computation are executed by Eigen library, which needs an Eigen::GpuDevice type object as parameter in a GPU card, and the GpuDevice parameter is constructed with an Eigen::CudaStreamDevice object. We need to set both a specific GpuID and CudaStream to create a Eigen::CudaStream object.
+
+At the same time, some computation work will executed by cuBLAS or cuDNN library. Take cuBLAS library as an example, we have to acquire a cublasHandle which binds on a CudaStream to make computation. It's the same way as Eigen library does.
+
+`DeviceContext` is defined as follows(just using for unify class `CudaDeviceContext` and class `CpuDeviceContext`):
```

Comment:
Yes, we need stream, cublasHandle, cudnnHandle in GPU, but nearly nothing in CPU.
So, CUPDeviceContext and CUDADeviceContext can hardly have something in common.
 The DeviceContext is just used for unify two types and make it convenient passing parameter to Operator.

In caffe2, except for memory New and Delete,```CUPContext``` and ```CUDAContext``` nearly have no methods in common. (https://github.com/caffe2/caffe2/blob/master/caffe2/core/context.h#L124)
In mxnet, ```Stream<cpu>``` and ```Stream<gpu>``` nearly have no methods in common too. The Stream in mxnet holds runtime resoureces. (https://github.com/dmlc/mshadow/blob/20b54f068c1035f0319fa5e5bbfb129c450a5256/mshadow/tensor.h#L370)